webui/react/src/services/stream/stream.test.ts

```diff
@@ -0,0 +1,192 @@
+import WS from 'jest-websocket-mock';
+
+import { ProjectSpec } from './projects';
+import { sleep, Stream } from './stream';
+
+const onUpsert = vi.fn();
+const onDelete = vi.fn();
+const isLoading = vi.fn();
+const url = 'ws://localhost:1234';
+const genServer = () => new WS(url, { jsonProtocol: true });
+
+const spec1ws = [1];
+const spec2ws = [1, 2];
+const spec1 = new ProjectSpec(spec1ws);
+const spec2 = new ProjectSpec(spec2ws);
+
+const setup = () => {
+  const server = genServer();
+  return { client: new Stream(url, onUpsert, onDelete, isLoading), server };
+};
+
+const cleanup = (arr: Array<WS | Stream>) => {
+  for (const a of arr) a.close();
+};
+
+describe('stream', () => {
+  afterEach(() => {
+    vi.clearAllMocks();
+  });
+  it('auto reconnect', async () => {
+    const { server, client } = setup();
+    let s = await server.connected;
+    expect(s.readyState).toEqual(1);
+
+    server.error();
+    await sleep(0.1);
+    const newServer = genServer();
+    s = await newServer.connected;
```

Comment:
why do you need a sleep here?  Why not just create the new server immediately and await it's connected state?

Reply:
This is not necessary, I was trying to mimic that the server was out of reach for "a while"

Reply:
My guess is that since the websocket API is event-based, that you don't get any additional code coverage by having a wait in the test, and therefore it's just a waste of time.

If I'm wrong, feel free to keep it.  But if all the client code goes through the exact same codepaths, I'd say it's best to drop the sleep.